### ISSUE TITLE
Twisted HTTP client timeouts can fail for TLS connections.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8318]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8318|
|Type|defect|
|Created|2016-05-05 12:00:41Z|
|Branch|https://github.com/twisted/twisted/tree/abort-oldclient-8318|

Found while attempting to remove HTTP pipelining: this bug causes test breakages.

Twisted's old HTTP client module has the method `downloadPage` and `HTTPDownloader` class,  both of which give users the ability to provide a `timeout` parameter which in theory limits the amount of time these objects can spend downloading the page in question.

Users would expect that when providing a timeout to `downloadPage`, after approximately the amount of time given from the `timeout` parameter the download will abort and the returned deferred will fire with an appropriate error. However, this does not always happen. In particular, when using TLS it is possible to arrange a situation where this deferred does not fire after the timeout, even though the download is aborted.

This occurs because `HTTPClientFactory` ensures that the first callback in the deferred chain is one that waits for disconnection before continuing. This means that if any scenario occurs whereby calling `loseConnection` (which is what the timeout does) does *not* cause connection teardown, the timeout will fire but the first callback will hold everything up until disconnection occurs.

This matters because for TLS connections, `loseConnection` attempts an orderly shutdown of the TLS connection: that is, it issues a TLS shutdown. It then waits for acknowledgement of the TLS shutdown before shutting down the underlying transport, which then notifies the `HTTPClientFactory` of the disconnect.

This means that if the TLS shutdown is ignored for any reason, the download deferred will hang, despite the timeout firing. This shutdown can be ignored if, for example, the remote peer is not prepared to write anything to the socket, for example, *or* if the remote peer is not reading from the socket. That is how I bumped into this issue: I had a TLS connection on which I'd called pauseProducing, causing Twisted to stop reading from the socket. This meant that Twisted did not spot the shutdown and did not respond to it, leading the connection to hang.

To reproduce this, here are two scripts. Here is the client:

```
from twisted.internet import reactor
from twisted.web import client

def printResult(result):
    print "Job complete!"
    print result
    reactor.stop()

d = client.downloadPage(
    "https://localhost:8081/",
    "out.txt",
    timeout=0.05
)
d.addBoth(printResult)

reactor.callLater(120, reactor.stop)
reactor.run()

```

Here is the server (requires Python 2.7.9 or later). The server script takes two CLI arguments: the first points to a TLS cert chain and the second to the key for that chain. The client does not appear to validate the cert, so any self-signed cert will do.

```
import socket
import ssl
import sys

CERT_LOCATION = sys.argv[1]
KEY_LOCATION = sys.argv[2]

ctx = ssl.create_default_context(purpose=ssl.Purpose.CLIENT_AUTH)
ctx.load_cert_chain(CERT_LOCATION, KEY_LOCATION)

s = socket.socket()
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind(('', 8081))
s.listen(5)

while True:
    new_sock, _ = s.accept()
    new_sock = ctx.wrap_socket(new_sock, server_side=True)

    while True:
        data = new_sock.recv(65535)
        if not data:
            break

```

To reproduce, run the server and then run the client.

Expected behaviour:

The client will exit almost immediately complaining about a timeout.

Actual behaviour:

The client hangs until either the script is killed or two minutes have passed, at which point it exits.

Attachments:

* [8318_1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb9/bb986bbeb15d91025728b12f033501a4894cb481/8edcc9e52cfe0cb8e3b98b42b46778a08fae7ed4.patch) (886 bytes) - added by Lukasa on 2016-05-06 08:34:52Z - First draft patch
* [8318_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bb9/bb986bbeb15d91025728b12f033501a4894cb481/0882c7a521afccc5208db097c3f3109ae74ec824.patch) (3227 bytes) - added by Lukasa on 2016-05-06 08:49:04Z - Same as the first patch, but with a test!

<details><summary>Searchable metadata</summary>

```
trac-id__8318 8318
type__defect defect
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__branches_abort_oldclient_8318 branches/abort-oldclient-8318
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1462449641574554 1462449641574554
changetime__1462714613104140 1462714613104140
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

Ok, so here's one way to solve this problem: don't `loseConnection`, `abortConnection` instead. This stops Twisted from waiting for the TLS response that may never be coming. It's dramatically less clean, but if the timeout has fired I don't think we care about how clean it is.

This is available for review now. I haven't written a test here because it's not immediately clear that we can nicely write one, but I'll have a think and see.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_2"></a>|@Lukasa commented|
|-|-|

Alright, I've updated this to add a test. The test 'fails' by timing out, which happily trial will eventually spot. I'm open to suggestions for how to make that test fail faster, but for now this works.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

(In [47353]) Branching to abort-oldclient-8318.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

(In [47354]) apply patch from lukasa, refs [#8318](https://github.com/twisted/twisted/issues/8318)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set owner to @hawkowl|
|-|-|

Hi Lukasa, thanks for the patch. The builders are good, I'll merge tomorrow morning. There's no better way of sorting out the old client code than this, so, I think we'll be good for now, it all needs to be deprecated anwyay.

Only minor thing; the topfile wasn't to spec (https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles) but I've fixed this.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl set status to `closed`|
|-|-|

(In [47359]) Merge abort-oldclient-8318: Make the old t.w.client APIs timeout correctly on TLS connections

Author: lukasa
Reviewer: hawkowl
Fixes: [#8318](https://github.com/twisted/twisted/issues/8318)
